modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/impl/VolatileLogStorageFactory.java

```diff
@@ -17,15 +17,64 @@
 
 package org.apache.ignite.internal.raft.storage.impl;
 
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.Set;
+import org.apache.ignite.configuration.schemas.table.LogStorageBudgetView;
 import org.apache.ignite.internal.raft.storage.LogStorageFactory;
+import org.apache.ignite.lang.IgniteInternalException;
+import org.apache.ignite.raft.jraft.core.LogStorageBudgetFactory;
+import org.apache.ignite.raft.jraft.core.LogStorageBudgetsModule;
 import org.apache.ignite.raft.jraft.option.RaftOptions;
 import org.apache.ignite.raft.jraft.storage.LogStorage;
 import org.apache.ignite.raft.jraft.storage.impl.LocalLogStorage;
+import org.apache.ignite.raft.jraft.storage.impl.LogStorageBudget;
+import org.apache.ignite.raft.jraft.storage.impl.VolatileLogStorage;
 
 /**
  * Log storage factory based on {@link LocalLogStorage}.
  */
 public class VolatileLogStorageFactory implements LogStorageFactory {
+    private final LogStorageBudgetView logStorageBudgetConfig;
+
+    private final Map<String, LogStorageBudgetFactory> budgetFactories;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param logStorageBudgetConfig Budget config.
+     */
+    public VolatileLogStorageFactory(LogStorageBudgetView logStorageBudgetConfig) {
+        this.logStorageBudgetConfig = logStorageBudgetConfig;
+
+        Map<String, LogStorageBudgetFactory> factories = new HashMap<>();
+
+        ClassLoader serviceClassLoader = Thread.currentThread().getContextClassLoader();
+
+        for (LogStorageBudgetsModule module : ServiceLoader.load(LogStorageBudgetsModule.class, serviceClassLoader)) {
+            Map<String, LogStorageBudgetFactory> factoriesFromModule = module.budgetFactories();
+
+            checkForBudgetNameClashes(factories.keySet(), factoriesFromModule.keySet());
+
+            factories.putAll(factoriesFromModule);
+        }
+
+        budgetFactories = Map.copyOf(factories);
+    }
+
+    private void checkForBudgetNameClashes(Set<String> names1, Set<String> names2) {
+        Set<String> intersection = new HashSet<>(names1);
+        intersection.retainAll(names2);
+
+        if (!intersection.isEmpty()) {
+            throw new IgniteInternalException(
```

Comment:
Is there any possibility to not use deprecated constructor for IgniteInternalException?
Maybe we should introduce some error codes for that scenario

Reply:
Added a TODO to address this later